src/lib.rs

```diff
@@ -98,16 +102,66 @@ impl Diagnostic {
         }
     }
 }
-
-pub type Sources<'a> = [&'a dyn SourceContainer];
+pub trait SourceLocation {
+    /// returns the location of this source-container. Used when reporting diagnostics.
+    fn get_location(&self) -> Option<&str>;
+}
 
 /// SourceContainers offer source-code to be compiled via the load_source function.
 /// Furthermore it offers a location-String used when reporting diagnostics.
-pub trait SourceContainer {
+pub trait SourceContainer: SourceLocation {
     /// loads and returns the SourceEntry that contains the SourceCode and the path it was loaded from
-    fn load_source(&self) -> Result<SourceCode, String>;
-    /// returns the location of this source-container. Used when reporting diagnostics.
-    fn get_location(&self) -> &str;
+    fn load_source(self, encoding: Option<&'static Encoding>) -> Result<SourceCode, String>;
+    // fn get_location(&self) -> &str;
```

Comment:
this comment stayed here because you wanted to think about joining the two traits again ;-)

Reply:
Indeed :)